File: App/NodeProcessors/StockNode.jl
module StockNode

using ..VectoriseBacktestService
using ..VectoriseBacktestService.Types
using ..VectoriseBacktestService.FlowData
using ..VectoriseBacktestService.ErrorHandlers

export process_stock_node

"""
Validates the stock node structure and properties
"""
function validate_stock_node(stock_node::Dict{String,Any})
    if !haskey(stock_node, "properties")
        throw(ValidationError("Missing properties in stock node"))
    end

    properties = stock_node["properties"]
    if !haskey(properties, "symbol")
        throw(ValidationError("Missing symbol in stock node properties"))
    end

    if !isa(properties["symbol"], String)
        throw(
            ValidationError(
                "Invalid symbol type",
                Dict("expected" => "String", "got" => typeof(properties["symbol"])),
            ),
        )
    end

    if isempty(properties["symbol"])
        throw(ValidationError("Empty symbol in stock node properties"))
    end
end

"""
Validates input parameters for stock node processing
"""
function validate_inputs(
    active_branch_mask::BitVector,
    total_days::Int,
    node_weight::Float32,
    portfolio_history::Vector{DayData},
)
    if isempty(active_branch_mask)
        throw(ValidationError("Empty active branch mask"))
    end

    if total_days <= 0
        throw(ValidationError("Invalid total days", total_days))
    end

    if node_weight <= 0 || node_weight > 1
        throw(ValidationError("Invalid node weight", node_weight))
    end

    if isempty(portfolio_history)
        throw(ValidationError("Empty portfolio history"))
    end

    if length(portfolio_history) < total_days
        throw(
            ValidationError(
                "Portfolio history shorter than total days",
                Dict(
                    "portfolio_length" => length(portfolio_history),
                    "total_days" => total_days,
                ),
            ),
        )
    end
end

"""
Updates portfolio with stock information
"""
function update_portfolio(
    portfolio_history::Vector{DayData},
    symbol::String,
    node_weight::Float32,
    active_days::Vector{Int},
    total_days::Int,
)
    try
        for day in active_days
            if day > length(portfolio_history) || day <= 0
                throw(
                    StockProcessingError(
                        "Invalid day index",
                        Dict("day" => day, "portfolio_length" => length(portfolio_history)),
                    ),
                )
            end

            portfolio_idx = length(portfolio_history) - total_days + day
            if portfolio_idx <= 0 || portfolio_idx > length(portfolio_history)
                throw(
                    StockProcessingError(
                        "Invalid portfolio index",
                        Dict(
                            "calculated_index" => portfolio_idx,
                            "portfolio_length" => length(portfolio_history),
                        ),
                    ),
                )
            end

            push!(
                portfolio_history[portfolio_idx].stockList, StockInfo(symbol, node_weight)
            )
        end
    catch e
        if e isa BacktestError
            rethrow(e)
        else
            throw(
                StockProcessingError(
                    "Failed to update portfolio", Dict("symbol" => symbol, "error" => e)
                ),
            )
        end
    end
end

"""
Processes a stock node in the trading strategy
"""
function process_stock_node(
    stock_node::Dict{String,Any},
    active_branch_mask::BitVector,
    total_days::Int,
    node_weight::Float32,
    portfolio_history::Vector{DayData},
    flow_count::Dict{String,Int},
    flow_stocks::Dict{String,Vector{DayData}},
)
    try
        # Validate inputs
        validate_stock_node(stock_node)
        validate_inputs(active_branch_mask, total_days, node_weight, portfolio_history)

        # Extract symbol
        symbol::String = stock_node["properties"]["symbol"]
        active_days::Vector{Int} = findall(active_branch_mask)

        # Update flow count if hash exists
        if haskey(stock_node, "hash")
            try
                increment_flow_count(flow_count, stock_node["hash"])
            catch e
                throw(
                    StockProcessingError(
                        "Failed to increment flow count",
                        Dict("hash" => stock_node["hash"], "error" => e),
                    ),
                )
            end
        end

        # Update portfolio
        update_portfolio(portfolio_history, symbol, node_weight, active_days, total_days)

        # Set flow stocks if hash exists
        if haskey(stock_node, "hash")
            try
                set_flow_stocks(flow_stocks, portfolio_history, stock_node["hash"])
            catch e
                throw(
                    StockProcessingError(
                        "Failed to set flow stocks",
                        Dict("hash" => stock_node["hash"], "error" => e),
                    ),
                )
            end
        end

        return total_days  # Successful processing
    catch e
        if e isa BacktestError
            rethrow(e)
        else
            throw(
                StockNodeError(
                    "Failed to process stock node",
                    Dict(
                        "symbol" =>
                            get(get(stock_node, "properties", Dict()), "symbol", "unknown"),
                        "node_id" => get(stock_node, "id", "unknown"),
                        "error" => e,
                    ),
                ),
            )
        end
    end
end

end

File: cpp_backtester/src/nodes/stock_node.cpp
#include \"stock_node.h\"
#include <algorithm>
#include <stdexcept>

namespace atlas {

NodeResult StockNodeProcessor::process(
    const StrategyNode& node,
    std::vector<bool>& active_mask,
    int total_days,
    float node_weight,
    std::vector<DayData>& portfolio_history,
    const std::vector<std::string>& date_range,
    std::unordered_map<std::string, int>& flow_count,
    std::unordered_map<std::string, std::vector<DayData>>& flow_stocks,
    std::unordered_map<std::string, std::vector<float>>& indicator_cache,
    std::unordered_map<std::string, std::vector<float>>& price_cache,
    const Strategy& strategy,
    bool live_execution,
    int global_cache_length
) {
    try {
        // Validate stock node
        if (!validate_stock_node(node)) {
            return NodeResult(0, false, \"Invalid stock node structure\");
        }
        
        // Validate inputs
        if (!validate_inputs(active_mask, total_days, node_weight, portfolio_history)) {
            return NodeResult(0, false, \"Invalid input parameters\");
        }
        
        // Extract symbol from properties
        std::string symbol = node.properties[\"symbol\"].get<std::string>();
        
        // Find active days
        auto active_days = find_active_days(active_mask);
        
        // Update flow count if hash exists
        if (!node.hash.empty()) {
            increment_flow_count(flow_count, node.hash);
        }
        
        // Update portfolio
        update_portfolio(portfolio_history, symbol, node_weight, active_days, total_days);
        
        // Set flow stocks if hash exists
        if (!node.hash.empty()) {
            set_flow_stocks(flow_stocks, portfolio_history, node.hash);
        }
        
        return NodeResult(total_days, true);
        
    } catch (const std::exception& e) {
        return NodeResult(0, false, \"Stock node processing error: \" + std::string(e.what()));
    }
}

bool StockNodeProcessor::validate_stock_node(const StrategyNode& node) const {
    // Must have properties
    if (node.properties.empty()) {
        return false;
    }
    
    // Must have symbol in properties
    if (!node.properties.contains(\"symbol\")) {
        return false;
    }
    
    // Symbol must be a non-empty string
    if (!node.properties[\"symbol\"].is_string()) {
        return false;
    }
    
    std::string symbol = node.properties[\"symbol\"].get<std::string>();
    if (symbol.empty()) {
        return false;
    }
    
    return true;
}

bool StockNodeProcessor::validate_inputs(
    const std::vector<bool>& active_mask,
    int total_days,
    float node_weight,
    const std::vector<DayData>& portfolio_history
) const {
    // Check active mask
    if (active_mask.empty()) {
        return false;
    }
    
    // Check total days
    if (total_days <= 0) {
        return false;
    }
    
    // Check node weight
    if (node_weight <= 0.0f || node_weight > 1.0f) {
        return false;
    }
    
    // Check portfolio history
    if (portfolio_history.empty()) {
        return false;
    }
    
    // Portfolio history must be at least as long as total days
    if (static_cast<int>(portfolio_history.size()) < total_days) {
        return false;
    }
    
    return true;
}

void StockNodeProcessor::update_portfolio(
    std::vector<DayData>& portfolio_history,
    const std::string& symbol,
    float node_weight,
    const std::vector<int>& active_days,
    int total_days
) {
    for (int day : active_days) {
        // Calculate portfolio index (Julia-style indexing from end)
        int portfolio_idx = static_cast<int>(portfolio_history.size()) - total_days + day;
        
        // Validate index
        if (portfolio_idx < 0 || portfolio_idx >= static_cast<int>(portfolio_history.size())) {
            throw std::runtime_error(\"Invalid portfolio index: \" + std::to_string(portfolio_idx));
        }
        
        // Add stock to portfolio
        StockInfo stock_info(symbol, node_weight);
        portfolio_history[portfolio_idx].add_stock(stock_info);
    }
}

std::vector<int> StockNodeProcessor::find_active_days(const std::vector<bool>& active_mask) const {
    std::vector<int> active_days;
    
    for (size_t i = 0; i < active_mask.size(); ++i) {
        if (active_mask[i]) {
            active_days.push_back(static_cast<int>(i));
        }
    }
    
    return active_days;
}

} // namespace atlas